sound/soc/sof/sof-audio.c

```diff
@@ -158,12 +203,25 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
 					 &r, sizeof(r));
 		break;
 	}
-	if (ret < 0)
-		dev_err(sdev->dev, "error: failed to load widget type %s\n",
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to load widget %s\n", swidget->widget->name);
+		goto use_count_free;
+	}
+
+	/* restore kcontrols for widget */
+	ret = sof_widget_kcontrol_setup(sdev, swidget);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to restore kcontrols for widget %s\n",
 			swidget->widget->name);
-	else
-		dev_dbg(sdev->dev, "widget %s setup complete\n", swidget->widget->name);
+		sof_widget_free(sdev, swidget);
```

Comment:
this really doesn't look right. Is there a real reason why in the error case above we only decrement the use-count and here we free the widget?